### PR TITLE
PIM-5594: Store panel state in session storage

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

Ensure panel state even after page changes. The panel will stay open if you leave the product page or you go to the next page (sequential edit).

Boyscout stuff:

- Move a behat scenario to the proper location
- Remove the listener on the state change (magical backbone stuff)
- Remove the multiple panel bug (multiple completeness, comments and history buttons)
- Added jsdoc to the panels.js file

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | NA
| Added Behats                      | Y
| Changelog updated                 | NA
| Review and 2 GTM                  | N
| Micro Demo to the PO (Story only) | N
| Migration script                  | NA
| Tech Doc                          | NA

### CODE REVIEW

File: features/product/ensure_panel_state.feature [hunk 1]
```diff
@@ -0,0 +1,24 @@
+@javascript
+Feature: Keep the side panel open in case between each product
```
Comment (phaseinducer) [inline features/product/ensure_panel_state.feature]: "in case" ?

File: features/product/ensure_panel_state.feature [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  In order to be productive with the pef
```
Comment (phaseinducer) [inline features/product/ensure_panel_state.feature]: *product edit form

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js [hunk 3]
```diff
@@ -23,65 +23,127 @@ define(
...
             configure: function () {
                 this.onExtensions('panel:register', this.registerPanel.bind(this));
                 this.listenTo(this.getRoot(), 'pim_enrich:form:render:after', this.resize);
 
                 return BaseForm.prototype.configure.apply(this, arguments);
             },
+
+            /**
+             * Register a panel to be displayed
+             * @param {Event} event
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: blank line please

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js [hunk 4]
```diff
@@ -23,65 +23,127 @@ define(
             events: {
                 'click > header > .close': 'closePanel'
             },
+
+            /**
+             * {@inheritdoc}
+             */
             initialize: function () {
                 this.state = new Backbone.Model();
```
Comment (jmleroux) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: Do we really need a Backbone.Model ?
Could not a plain vanilla object be sufficient ?

File: features/product/ensure_panel_state.feature (hunk 1 above)
Comment (grena) [inline features/product/ensure_panel_state.feature]: > Keep the side panel open when switching product page

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js [hunk 5]
```diff
@@ -23,65 +23,127 @@ define(
...
+             */
             closeFullPanel: function () {
                 this.getParent().setFullPanel(false);
             },
+
+            /**
+             * Get the curent panel code
+             */
+            getCurrentPanel: function () {
+                return sessionStorage.getItem('current_form_panel_' + this.getFormCode());
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: What do you think about something like `this.getFormCode() + '.current_panel'` ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js [hunk 6]
```diff
@@ -23,65 +23,127 @@ define(
...
             render: function () {
                 if (!this.configured) {
                     return this;
                 }
 
-                this.$el[this.state.get('currentPanel') ? 'removeClass' : 'addClass']('closed');
+                this.$el[this.getCurrentPanel() ? 'removeClass' : 'addClass']('closed');
 
-                var currentPanel = _.findWhere(this.state.get('panels'), {code: this.state.get('currentPanel')});
+                var currentPanel = this.getPanel(this.getCurrentPanel());
```
Comment (jmleroux) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: a bit weird to read, no ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js [hunk 7]
```diff
@@ -23,65 +23,127 @@ define(
...
+                this.render();
+            },
+
+            /**
+             * Get the panel data from code
+             * @param {String}
+             *
+             * @return {Object}
+             */
+            getPanel: function (code) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: Too ambiguous with the introduction of new methods. `getPanelFromCode` maybe ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js (hunk 6 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: This is weird, either `this.getPanel` is useless, or `this.getCurrentPanel()` should be called `this.getCurrentPanelCode()`

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js [hunk 8]
```diff
@@ -23,65 +23,127 @@ define(
...
+             * Close the full size panel
+             */
             closeFullPanel: function () {
                 this.getParent().setFullPanel(false);
             },
+
+            /**
+             * Get the curent panel code
+             */
+            getCurrentPanel: function () {
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: `getCurrentPanelCode`

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js [hunk 9]
```diff
@@ -23,65 +23,127 @@ define(
...
+                return sessionStorage.getItem('current_form_panel_' + this.getFormCode());
+            },
+
+            /**
+             * Set the curent panel code
+             */
+            setCurrentPanel: function (code) {
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: `setCurrentPanelCode`

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js (hunk 7 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: :+1:

**Comment by grena** [comment]
GTM once comments fixed

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/selector.js [hunk 10]
```diff
@@ -23,7 +23,8 @@ define(
             render: function () {
                 this.$el.html(
                     this.template({
-                        state: this.getParent().state.toJSON()
+                        panels: this.getParent().state.get('panels'),
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/selector.js]: We should do a getter for that.

**Comment by jmleroux** [comment]
GTM once no more ambiguity between `getCurrentPanel` and `getPanel`

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js (hunk 4 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: yup, I fix that

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js (hunk 5 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: no really difference, I just followed the method we use elsewhere

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js (hunk 6 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: If `getPanel()` is never used for somehting else than the current panel we should drop it yes.
#yoloJsBcBreak

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/product/panel/selector.html [hunk 11]
```diff
@@ -1,3 +1,3 @@
-<% _.each(state.panels, function(panel) { %>
-    <button class="btn <%- panel.code === state.currentPanel ? 'active' : '' %>" data-panel="<%- panel.code %>"><%- panel.label %></button>
+<% _.each(panels, function(panel) { %>
+    <button class="btn <%- panel.code === currentPanel ? 'active' : '' %>" data-panel="<%- panel.code %>"><%- panel.label %></button>
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/product/panel/selector.html]: Way better to not receive the state^^ :+1:

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js (hunk 5 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/panel/panels.js]: Ok, just thought of something more readable.

**Pull request merged**